### PR TITLE
[ADD] 8.0 - New module 'base_user_role' to manage user roles efficiently

### PR DESCRIPTION
**Pull request by sebalix**
Note : we are using this module on version 6.0, 7.0 and 8.0 I just added the support for dates recently on the 8.0 version. I'll migrate it to 9.0 and 10.0 soon.

Related issue: #146 

User roles
========

This module was written to extend the standard functionality regarding users and groups management. It helps creating well-defined user roles and associating them to users.

It can become very hard to maintain a large number of user profiles over time, juggling with many technical groups. For this purpose, this module will help you to:

  * define functional roles by aggregating low-level groups,
  * set user accounts with the predefined roles (roles are cumulative),
  * update groups of all relevant user accounts (all at once),
  * ensure that user accounts will have the groups defined in their roles (nothing more, nothing less). In other words, you can not set groups manually on a user as long as there is roles configured on it,
  * activate/deactivate roles depending on the date (useful to plan holidays, etc)
  * get a quick overview of roles and the related user accounts.

That way you make clear the different responsabilities within a company, and are able to add and update user accounts in a scalable and reliable way.

Configuration
=============

To configure this module, you need to go to *Configuration / Users / Roles*, and create a new role. From there, you can add groups to compose your role, and then associate users to it.

### CODE REVIEW

**Comment by sebalix** [comment]
About Travis: flake8 errors and tests failure not related to this module.

**Comment by jcdrubay** [comment]
Very good contribution (high quality of code, useful, well documented).
I hope it can be reviewed and merged soon :+1:

File: base_user_role/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+{
+    'name': 'User roles',
+    'version': '8.0.1.1.0',
+    'category': 'Tools',
+    'author': 'ABF OSIELL, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'maintainer': 'ABF OSIELL',
+    'website': 'http://www.osiell.com',
+    'depends': [
```
Comment (lasley) [inline base_user_role/__openerp__.py:13]: You can remove this key. `base` dependency is implicit.

File: base_user_role/models/role.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    group_id = fields.Many2one(
+        'res.groups', required=True, ondelete='cascade',
+        readonly=True, string=u"Associated group")
+    line_ids = fields.One2many(
+        'res.users.role.line', 'role_id', string=u"Users")
+    user_ids = fields.One2many(
+        'res.users', string=u"Users", compute='_compute_user_ids')
+
+    @api.multi
+    @api.depends('line_ids')
```
Comment (lasley) [inline base_user_role/models/role.py]: `line_ids.user_id`

File: base_user_role/models/role.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    def write(self, vals):
+        res = super(ResUsersRole, self).write(vals)
+        self.update_users()
+        return res
+
+    @api.multi
+    def update_users(self):
+        """Update all the users concerned by the roles identified by `ids`."""
+        users = self.mapped('user_ids')
+        if users:
```
Comment (lasley) [inline base_user_role/models/role.py]: There shouldn't be need for this guard, mapped will just return an empty Recordset that should work with the other method just fine due to record loop

File: base_user_role/models/role.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    _name = 'res.users.role.line'
+    _description = 'Users associated to a role'
+
+    role_id = fields.Many2one(
+        'res.users.role', string=u"Role")
+    user_id = fields.Many2one(
+        'res.users', string=u"User")
+    date_from = fields.Date(u"From")
+    date_to = fields.Date(u"To")
+    is_active = fields.Boolean(u"Active", compute='_compute_is_active')
```
Comment (lasley) [inline base_user_role/models/role.py]: `active` is the Odoo norm for this col - although I'm with you on prefixing bool cols with linking verbs like `is`, `has`, etc. It also comes with some defaults filters and such, so makes it worth using despite the hurt to naming OCD

That said, it also needs to be `store=True` to allow all of that

File: base_user_role/models/user.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+class ResUsers(models.Model):
+    _inherit = 'res.users'
+
+    role_line_ids = fields.One2many(
+        'res.users.role.line', 'user_id', string=u"Role lines")
+    role_ids = fields.One2many(
+        'res.users.role', string=u"Roles", compute='_compute_role_ids')
+
+    @api.multi
+    @api.depends('role_line_ids')
```
Comment (lasley) [inline base_user_role/models/user.py]: role_line_ids.role_id

**Review by lasley — CHANGES_REQUESTED** [review]
This is awesome, thanks @sebalix 

Code, no test. Comments inline

**Comment by lasley** [comment]
Oh please also check Travis and Runbot 😉

File: base_user_role/security/ir.model.access.csv [hunk 6]
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_res_users_role,access_res_users_role,model_res_users_role,"base.group_erp_manager",1,1,1,1
```
Comment (jcdrubay) [inline base_user_role/security/ir.model.access.csv:2]: From travis https://travis-ci.org/OCA/server-tools/jobs/176051212#L1703

> The model res.users.role.line has no access rules, consider adding one. E.g. 

`access_res_users_role_line,access_res_users_role_line,model_res_users_role_line,"base.group_erp_manager",1,0,0,0`

**Comment by sebalix** [comment]
Thank you for the review, I'll fix these points soon

File: base_user_role/__openerp__.py (hunk 1 above)
Comment (pedrobaeza) [reply base_user_role/__openerp__.py:13]: No, you should depend on your module at least in base, but base is implicit when depending on other module.

File: base_user_role/models/role.py (hunk 4 above)
Comment (sebalix) [reply base_user_role/models/role.py]: Indeed, but I was unable to set a working `domain` on the XML views to display all the lines (as we don't want to hide them from the administrator in any case, we just want to filter them when setting groups). If I set the domain on the Python field, it works fine, but then it does not dispense to filter the lines in the `ResUsers.set_groups_from_roles` method. And as I've seen no benefit to use the standard `active` field I ended up adding my own.

File: base_user_role/security/ir.model.access.csv (hunk 6 above)
Comment (sebalix) [reply base_user_role/security/ir.model.access.csv:2]: Fixed

File: base_user_role/models/role.py (hunk 4 above)
Comment (sebalix) [reply base_user_role/models/role.py]: Or I could rename it to 'is_enabled'?

File: base_user_role/models/role.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+class ResUsersRoleLine(models.Model):
+    _name = 'res.users.role.line'
+    _description = 'Users associated to a role'
+
+    role_id = fields.Many2one(
+        'res.users.role', string=u"Role", ondelete='cascade')
+    user_id = fields.Many2one(
+        'res.users', string=u"User")
+    date_from = fields.Date(u"From")
```
Comment (susport) [inline base_user_role/models/role.py:66]: Date from may be default the date now of the create moment.
Good contribution.
👍

Comment (sebalix) [reply base_user_role/models/role.py:66]: Dates are optional, no dates = role activated immediately

File: base_user_role/models/role.py (hunk 4 above)
Comment (lasley) [reply base_user_role/models/role.py]: I like the `is_enabled` idea - it is far enough away from `active` that people won't be trying to change it all the time. I know I would have probably done this during a migration had I not already gotten the reasoning behind it.

Comment (sebalix) [reply base_user_role/models/role.py]: It's done!

**Comment by sebalix** [comment]
I think all the points are fixed.

**Comment by lasley** [comment]
Travis is still mad 🔴

**Comment by sebalix** [comment]
You mean I have to fix lint errors of others modules in this one? Tests fail on `fetchmail_notify_error_to_sender`.

**Review by lasley — APPROVED** [review]
Hmm yeah ok the build looks broken on 8.0 branch in the same way, so you are correct that we should not be fixing it here.

**Comment by sebalix** [comment]
I fixed the other modules in #616 to get green builds.

**Comment by sebalix** [comment]
Rebased

**Comment by sbidoul** [comment]
@adrienpeiffer you should have a look at this.

**Comment by andhit-r** [comment]
Tested functionality on runbot :+1:

**Comment by andhit-r** [comment]
@mikevhe18 @azmimr67 you guys should take a look at this PR and help review

**Comment by lasley** [comment]
We're approved for merge. @sebalix - do you prefer to squash your own commits or would you rather me do it? I don't think I've gotten this preference from you yet, but sorry if I'm double asking.

**Comment by sebalix** [comment]
@lasley You can squash them, no trouble

**Comment by lasley** [comment]
Thanks for the awesome new module @sebalix! Do you have any plans to migrate it forward?

**Comment by sebalix** [comment]
@lasley Thank you! Yes, I plan to migrate it to v9 (see #641) and v10 right after.

**Pull request merged**